### PR TITLE
Added travis task for functional tests. Js error handling tests added (closes #291)

### LINKED ISSUE #291: Add 'Uncaught js error handling' functional tests

### PR DESCRIPTION
**Pull request by georgiy-abbasov**

### CODE REVIEW

**Comment by inikulin** [comment]
Guys let's switch to the node:stable for qunit and functional tests in this PR

**Comment by inikulin** [comment]
I wonder it took so long to accomplish tests. Each test takes ~5s. There are 12 tests, so it should take ~60s. Meanwhile cumulative time is 3m. \cc @AlexanderMoskovkin @georgiy-abbasov

**Comment by georgiy-abbasov** [comment]
@inikulin, it takes about 2 minutes to open SauceLabs tunnel and start 5 virtual machines. These actions take place in setup.js, i.e. mocha 'before all' method.

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot /retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
Let's make sure that tests are stable

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
lgtm 
only let @VasilyStrelyaev check strings

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@VasilyStrelyaev, @inikulin
fpr

File: test/functional/setup.js [hunk 1]
```diff
@@ -91,9 +91,10 @@ before(function () {
         })
         .then(function () {
             global.runTests = function (fixture, testName, opts) {
-                var report = '';
+                var report       = '';
                 var runner = testCafe.createRunner();
                 var fixturePath = path.join(path.dirname(caller()), fixture);
+                var skipJsErrors = opts && opts.skipJsErrors;
```
Comment (inikulin) [inline test/functional/setup.js]: Incorrect var declaration formatting. Please, setup WS formatting according to our rules

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
fpr

File: test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
+var expect        = require('chai').expect;
+var customJsError = '%TC_TEST_ERR%';
+
+describe('Uncaught js errors', function () {
+    it('Should fail when there are no onerror handler', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js]: it('Should fail when there is no onerror handler', function () {

File: test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    it('Should fail when iframe\'s onerror handler returns undefined', function () {
+        return runTests('same-domain-iframe.test.js', null, { shouldFail: true })
+            .catch(function (err) {
+                expect(err).to.contains(customJsError);
+            });
+    });
+
+    it('Should fail when loaded page throws error', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js]:
```
    it('Should fail when a loaded page throws an error', function () {
```

File: test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        return runTests('loaded.test.js', null, { shouldFail: true })
+            .catch(function (err) {
+                expect(err).to.contains(customJsError);
+            });
+    });
+
+    it('Should success when the error raised in cross-domain iframe', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js]:
```
    it('Should success when an error is raised in a cross-domain iframe', function () {
```

File: test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    it('Should success when onerror handler returns true', function () {
+        return runTests('handler-returns-true.test.js', null)
+            .then(function (err) {
+                expect(err).eql('');
+            });
+    });
+
+    it('Should fail when there are no onerror handler (skipJsErrors enabled)', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js]:
```
    it('Should fail when there is no onerror handler (skipJsErrors enabled)', function () {
```

File: test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    it('Should success when the error raised in cross-domain iframe', function () {
+        return runTests('cross-domain-iframe.test.js', null)
+            .then(function (err) {
+                expect(err).eql('');
+            });
+    });
+
+    it('Should success when onerror handler returns true', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js]:
```
    it('Should success when the onerror handler returns true', function () {
```

File: test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    it('Should fail when iframe\'s onerror handler returns undefined (skipJsErrors enabled)', function () {
+        return runTests('same-domain-iframe.test.js', null, { skipJsErrors: true })
+            .then(function (err) {
+                expect(err).eql('');
+            });
+    });
+
+    it('Should success when loaded page throws error (skipJsErrors enabled)', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js]:
```
    it('Should success when the loaded page throws an error (skipJsErrors enabled)', function () {
```

File: test/functional/fixtures/runner/uncaught-js-errors/cross-domain-iframe.test.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+'@fixture cross domain iframe';
+'@page ./cross-domain-iframe.html';
+
+'@test'['cross domain iframe test'] = {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/runner/uncaught-js-errors/cross-domain-iframe.test.js]: '@test'['cross-domain iframe test'] = {

File: test/functional/fixtures/runner/uncaught-js-errors/same-domain-iframe.test.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+'@fixture same domain iframe';
+'@page ./same-domain-iframe.html';
+
+'@test'['same domain iframe test'] = {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/runner/uncaught-js-errors/same-domain-iframe.test.js]:
```
'@test'['same-domain iframe test'] = {
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
fpr

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by inikulin** [comment]
lgtm

**Comment by inikulin** [comment]
We need build-matrix for this in readme I guess. \cc @AlexanderMoskovkin @georgiy-abbasov

**Comment by georgiy-abbasov** [comment]
@inikulin, i'll do this.

**Comment by AlexanderMoskovkin** [comment]
@inikulin it's not so trivial as with `qunit` tests, so it'll take some time

**Pull request merged**